enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,312 @@
+with the name (DNS or IP) and port of the cluster. As seen in previous section,
+if its a name it will find a matching certificate, and if its an IP it will try
+looking it up in the mapping table, returning the default certificate if not
+present. Note the name or IP must be in the external certificate, as there is
+hostname validation on the client.
+
+Internal access is typically using [client-go](https://github.com/kubernetes/client-go).
+If using [InClusterConfig](https://github.com/kubernetes/client-go/blob/master/rest/config.go#L511)
+then it will not use the name for connecting. As we can see it uses the IP for
+the `kubernetes` service, injected through the environment variables. This IP
+is the service network IP, which later on gets translated to an endpoint:
+```bash
+$ KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get svc kubernetes -o yaml
+apiVersion: v1
+kind: Service
+metadata:
+  creationTimestamp: "2022-12-20T15:57:23Z"
+  labels:
+    component: apiserver
+    provider: kubernetes
+  name: kubernetes
+  namespace: default
+  resourceVersion: "199"
+  uid: 28e299a6-1578-4fd0-babd-660a1a86cdfb
+spec:
+  clusterIP: 10.43.0.1
+  clusterIPs:
+  - 10.43.0.1
+  internalTrafficPolicy: Cluster
+  ipFamilies:
+  - IPv4
+  ipFamilyPolicy: SingleStack
+  ports:
+  - name: https
+    port: 443
+    protocol: TCP
+    targetPort: 6443
+  sessionAffinity: None
+  type: ClusterIP
+status:
+  loadBalancer: {}
+
+$ KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get endpoints kubernetes -o yaml
+apiVersion: v1
+kind: Endpoints
+metadata:
+  creationTimestamp: "2022-12-20T15:57:23Z"
+  labels:
+    endpointslice.kubernetes.io/skip-mirror: "true"
+  name: kubernetes
+  namespace: default
+  resourceVersion: "201"
+  uid: 2c3d37b8-9e04-4779-9947-6230450e2336
+subsets:
+- addresses:
+  - ip: 192.168.122.117
+  ports:
+  - name: https
+    port: 6443
+    protocol: TCP
+```
+As we can see the endpoint is the node IP, which means that any connection to
+the service IP will get translated to the node IP. When reaching API server the
+destination IP will be that of the node, and if there is no match for that in
+the map the default certificate is returned. This works because the service
+network certificate includes the service IP for hostname validation.
+
+#### Including the IP in the certificates
+We have already seen that internal access always uses the node IP because of
+the endpoints.
+
+If we were to include the node IP in the service network certificate we would
+get an immediate match in the mapping for pods, and the hostname verification
+matches the service network IP. If the connection using the node IP as
+destination does not come from a pod we run into trouble. The service network
+certificate is signed by its own specific CA, so any external connection using
+the IP will get the internal certificate and wont validate against the external
+CA.
```

Comment:
@dhellmann - does openshift provide any mechanism for customers to provide a CA that will be used to sign all of the certs & CS used in a cluster, or is openshift hard-coded to use a self-signed root-of-trust?  It seems like having a real root-of-trust would simplify this problem and be more secure (although I suppose CA/cert/secret rotation would be a problem if nobody has worked that out).

Reply:
Did a little digging - looks like OCP's root-of-trust[ is the bootstrap cert](https://docs.openshift.com/container-platform/4.12/security/certificate_types_descriptions/bootstrap-certificates.html). It's good for 10 years, which is probably why I don't see anything about rotating it. Also don't see anything about providing an alternative root-of-trust.

Reply:
That matches what I remember from discussions with @stlaz in the past. 

Unless we need to deal with user-provided CAs as part of this, let's leave that for later.